feat: Improve state key generation for dashboards and charts

michael-s-molina opened the pull request:
### SUMMARY
When opening a dashboard or a chart in Explore, the URL will update with a new state key every time the page is reloaded even when the state isn't updated. This may not be scalable in high-traffic Superset deployments where it can quickly fill up disk space or exhaust the number of cache keys too fast, due to memory restrictions, making it more likely that a link you shared with someone a while ago becomes invalid.

To address this problem we should consider the request context such as user ID, resource ID, etc, to generate new keys only when the context changes. This way we can reuse existing keys but at the same time maintain the uniqueness of storage spaces for different users/dashboards/datasets/charts.

We could use HMAC to generate the same key for the same context but that would require key rotation management.

We can’t base key generation on previous keys because we have POST scenarios where the key is not present. One example would be when the user is inside a dashboard and selects “View chart in Explore”.

We can’t change the structure of the cache key to use the user and resource because this information is not present in the GET/DELETE requests and it would also require keeping backward compatibility with previous versions.

To implement the desired behavior we’ll keep a secondary mapping of user and resources with the previously generated keys. We'll store this mapping together with the entry so they’ll share the same expiration time. We’ll also keep it in sync with DELETE events. That way we can reuse the keys and keep the solution backward compatible.

Hat tip to @zhaoyongjie and @ktmud for bringing this concern during the reviews 👏🏼 

Follow-up of https://github.com/apache/superset/pull/18181

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/152404373-a9680644-402c-4e48-a155-b47881f13993.mov

### TESTING INSTRUCTIONS
1 - Check that `form_data_key` is reused when the same user is accessing the same resource (dashboard, dataset, or chart)
2 - Check that `form_data_key` changes when the user, dashboard, dataset, or chart changes.
3 - Check that the values associated with a given key are correctly retrieved from the server

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
feat: Improve state key generation for dashboards and charts

michael-s-molina committed:
Removes unused import

michael-s-molina committed:
Syncs the mapping on delete

michael-s-molina committed:
Fixes Cypress test

michael-s-molina commented:
> Question: I tried add a new chart to the dashboard but the form_data_key did not change, is that expected?

Yes. It's expected. The `native_filter_key` should be unique for the same dashboard and user.

michael-s-molina committed:
Fixes URL publishing

michael-s-molina commented:
> Hey @michael-s-molina it took a few clicks to reproduce the problem sorry about that, but as you will see in the video after going back and forth a few times the Dashboard Video Game Sales went from using `1BoGOSkIZ4VMAxBajAbHIGGewJ9lWQclxJlI1vtXT59ALa2LlvM_qXzvD27xDhDZ` to using `75_Q5ExSV_9F_hhG-ytrT65ohOCgIqPNF4M6Tc-XcRQEmBx1_y7n368eHFwQBtXu` which is also the same that USA Birth Names is using.

Thanks for spotting that @geido! I submitted a fix.

michael-s-molina committed:
Fixes Cypress test

villebro commented on superset/dashboards/filter_state/commands/create.py:
```diff
@@ -23,13 +25,17 @@
+        contextual_key = cache_key(actor.get_user_id(), resource_id)
+        key = cache_manager.filter_state_cache.get(contextual_key)
+        if not key:
+            key = token_urlsafe(48)
```
Not introduced here, but noticed this only now.. Since we're using the same `token_urlsafe(48)` both in dashboard and explore form data commands, could we DRY this by creating a util under the `key_value` package that hides both the token generator and the magic number? Something like

```python
def get_key() -> str:
    return token_urlsafe(48)
```

michael-s-molina replied:
Nice catch! Done.

villebro commented on superset/dashboards/filter_state/commands/create.py:
```diff
@@ -23,13 +25,17 @@
         resource_id = cmd_params.resource_id
         actor = cmd_params.actor
-        key = cache_key(resource_id, cmd_params.key)
+        contextual_key = cache_key(actor.get_user_id(), resource_id)
+        key = cache_manager.filter_state_cache.get(contextual_key)
+        if not key:
+            key = token_urlsafe(48)
```
This part I find problematic. I get the idea that we want to avoid creating new keys every time we load the same resource with the same key. But recycling the same key every time the same user navigates to the same resource is IMO slightly risky. I'm sure it's a fairly common use case to be exploring the same resource in multiple different tabs. Making them all share the same key-value pair will make it impossible to do this type of exploration without leaking state between the tabs. I tried to record a video to demonstrate what I mean, but here's the repro steps:
1. Open Superset in two tabs
2. Navigate the first tab to your favourite dashboard and set a few native filters
3. Navigate the second tab to the same dashboard
4. refresh the first tab and notice that the native filter state has been lost

michael-s-molina replied:
Nice point. I'll work on it and submit a new commit supporting this.

villebro reviewed (commented):
I agree it may be pragmatic to not initialize a new key-value pair every time the resource is reloaded with the same key. However, removing support for having separate state in two tabs on the same context (user + resource) is a problem. I would like to suggest a compromise: when opening the same resource _without_ a key, a new key-value pair is initialized. However, when the same user reloads the same resource _with_ the same key, the key would not change.

zhaoyongjie commented on superset/explore/form_data/commands/create.py:
```diff
@@ -40,7 +41,10 @@ def run(self) -> str:
             form_data = self._cmd_params.form_data
             check_access(dataset_id, chart_id, actor)
-            key = token_urlsafe(48)
+            contextual_key = cache_key(actor.get_user_id(), dataset_id, chart_id)
```
1. In order to separate different session for same user, use session id instead of user id
```python
from flask import session
...
...
            contextual_key = cache_key(session.get('_id'), dataset_id, chart_id)
```

2. there is already a general hash module in `superset.utils.hashing`
```
import from md5_sha_from_str from superset.utils.hashing

contextual_key = md5_sha_from_str(<string variable>)
```

michael-s-molina replied:
1. Great call on the different sessions approach! I modified the code to follow your suggestion.
2. I think we don't need the extra overhead of a hashing function. The concatenation is just fine for this use case. Even if we would apply a hashing function we would still need to concatenate the arguments. I left this untouched. Let me know if you disagree 😉

zhaoyongjie replied:
The objective is the same: get a unique cache. Thanks for the improvement.

zhaoyongjie commented on superset/dashboards/filter_state/commands/delete.py:
```diff
@@ -34,5 +34,7 @@ def delete(self, cmd_params: CommandParameters) -> bool:
             if entry:
                 if entry["owner"] != actor.get_user_id():
                     raise KeyValueAccessDeniedError()
+                contextual_key = cache_key(actor.get_user_id(), resource_id)
```
same above(commands/create.py)

zhaoyongjie commented on superset/dashboards/filter_state/commands/create.py:
```diff
@@ -23,13 +25,17 @@
         resource_id = cmd_params.resource_id
         actor = cmd_params.actor
-        key = cache_key(resource_id, cmd_params.key)
+        contextual_key = cache_key(actor.get_user_id(), resource_id)
```
same above(commands/create.py)

zhaoyongjie commented on superset/explore/form_data/commands/create.py:
```diff
@@ -49,6 +53,7 @@ def run(self) -> str:
                 cache_manager.explore_form_data_cache.set(key, state)
+                cache_manager.explore_form_data_cache.set(contextual_key, key)
             return key
         except SQLAlchemyError as ex:
```
There isn't SQLAlchemy operation, please change to a more explicit exception.

michael-s-molina replied:
The `check_access` function can emit SQLAlchemy errors from different functions. I'm just catching here to unify the exception type.

zhaoyongjie replied:
okay. got it.

zhaoyongjie commented on superset/key_value/api.py:
```diff
@@ -122,9 +113,7 @@ def put(self, pk: int, key: str) -> Response:
-            )
+            args = CommandParameters(actor=g.user, resource_id=pk, key=key)
             value = self.get_get_command()(args).run()
             if not value:
```
What happens if the cache temporarily unavailable? we should set cache if cache broke down.

michael-s-molina replied:
Great call! I handled this problem in https://github.com/apache/superset/pull/18624

zhaoyongjie commented:
@michael-s-molina Thanks for working on this! there are 2 concerns in this PR.
1. How to treat cache temporarily unavailable
a) make a chart
b) remove all cache data from Redis or filesystem(depend on how to set up Superset)
c) refresh explore page
d) chart will be crash

![image](https://user-images.githubusercontent.com/2016594/152942709-0c03831e-dc37-4660-83e8-295b30917454.png)

2. How to treat the same user login in different terminal
a) open same chart from 2 different browser, like Chrome and Firefox.
b) user have different session, but they can sync form data because they use same cache key. we can use session id to avoid it.

rusackas commented on superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx:
```diff
@@ -30,7 +30,7 @@ import { FeatureFlag, isFeatureEnabled } from 'src/featureFlags';
 import { DataMaskStateWithId, DataMaskWithId } from 'src/dataMask/types';
 import { useImmer } from 'use-immer';
-import { isEmpty, isEqual } from 'lodash';
+import { isEmpty, isEqual, debounce } from 'lodash';
```
Lodash's `debounce` is the debounce method itself. `SLOW_DEBOUNCE` is just a constant. We'll probably want to use both together

michael-s-molina replied:
Thanks for pointing `SLOW_DEBOUNCE` @zhaoyongjie!

michael-s-molina committed:
Adds multi session support

michael-s-molina commented:
@zhaoyongjie Thanks for bringing the multi-session and key expiration problems to my attention. I addressed both problems here and in https://github.com/apache/superset/pull/18624

zhaoyongjie commented:
@michael-s-molina In order to avoid code conflicts, we need to merge [PR-18624](https://github.com/apache/superset/pull/18624) first.

michael-s-molina committed:
Adds multi tab support for dashboards

michael-s-molina commented:
@villebro @zhaoyongjie I rebased the code and worked on supporting multi-tab states for the dashboards. The idea is to add an optional parameter called `tab_id` to the `filter_state` enpoint. The server will generate different keys for different tabs. On the client-side, I'm using a mix of the [Local Storage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) with the [Broadcast Channel API](https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API) to generate the tab ids. Unfortunately, Safari still does not support the native API (will do in a month), so I'm using [broadcast-channel](https://github.com/pubkey/broadcast-channel) for now. As soon as Safari 15.4 is released we can remove the dependency. I'll work on the Explore part now to complete the PR.

michael-s-molina committed:
Removes unused useRef

villebro reviewed (approved):
This is really impressive work - logic looks solid and works just as expected. Bravo! 👏

michael-s-molina committed:
Adds multi tab support to Explore

michael-s-molina committed:
Formats Python files

michael-s-molina commented:
@zhaoyongjie @villebro I added the Explore part. Now the PR is complete. In this part, I only replicated the server-side logic in Explore and extracted the `tab_id` handling to a custom hook so we can use it anywhere.

michael-s-molina committed:
Preserves the URL when opening on Explore

michael-s-molina committed:
Wait for tab_id to publish URL on Explore

zhaoyongjie commented:
Hi @michael-s-molina, Thanks for providing different tabs cache mechanisms.

The current solution causes the same session to have different cache keys due to different tab keys. Imagine if the user created 5 tabs, the cache system has to create 5 cache records. For this high-frequency operation, do we consider using some data structures to solve it? for example:
```
form_data_key = {
  'tab-key1': value,
  'tab-key2: value,
}
```

Second, could we consider using `window.sessionStorage` to preserve `tab-id`? if we use `sessionStorage` and `tab-key in form_data_key`, we probably don't need to introduce a new dependence `broadcast-channel` in frontend.

In the future, we can also improve `for_data_key = {tab-key: value}` into `for_data_key = LRU_ordered_dict` to optimize memory space.

michael-s-molina commented:
> The current solution causes the same session to have different cache keys due to different tab keys. Imagine if the user created 5 tabs, the cache system has to create 5 cache records. For this high-frequency operation, do we consider using some data structures to solve it?
> In the future, we can also improve `for_data_key = {tab-key: value}` into `for_data_key = LRU_ordered_dict` to optimize memory space.

I think this could be a further optimization that we can implement in the future 👍🏼 

> Second, could we consider using `window.sessionStorage` to preserve `tab-id`? if we use `sessionStorage` and `tab-key in form_data_key`, we probably don't need to introduce a new dependence `broadcast-channel` in frontend.

I considered this option but discarded it because of the following in [window.sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) docs:

> Duplicating a tab copies the tab's sessionStorage into the new tab.

michael-s-molina committed:
Removes tests

zhaoyongjie commented:
> > The current solution causes the same session to have different cache keys due to different tab keys. Imagine if the user created 5 tabs, the cache system has to create 5 cache records. For this high-frequency operation, do we consider using some data structures to solve it?
> > In the future, we can also improve `for_data_key = {tab-key: value}` into `for_data_key = LRU_ordered_dict` to optimize memory space.
> 
> I think this could be a further optimization that we can implement in the future 👍🏼
> 

Server-side save tab_id  can avoid client tab collision and reduce complexity. 


> > Second, could we consider using `window.sessionStorage` to preserve `tab-id`? if we use `sessionStorage` and `tab-key in form_data_key`, we probably don't need to introduce a new dependence `broadcast-channel` in frontend.
> 
> I considered this option but discarded it because of the following in [window.sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) docs:
> 
> > Duplicating a tab copies the tab's sessionStorage into the new tab.

Thanks for the explanation! `sessionStorage` is more suitable for this scenario.

michael-s-molina committed:
Restores test

michael-s-molina committed:
Mocks useTabId hook

michael-s-molina committed:
Changes the mock description

rosemarie-chiu commented:
🏷 2022.7

rosemarie-chiu commented:
🏷 preset:2022.7
